#14033: Adding mypy type annotations to stats_domain.py

winnie368c opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of #14033.
2. This PR does the following: adds mypy type annotations to stats_domain.py

## Essential Checklist

- [X] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [X] The linter/Karma presubmit checks have passed locally on your machine.
- [X] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [X] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct
![allcheckspassed](https://user-images.githubusercontent.com/85202273/168954388-085ef2c4-f6bc-4c94-bcfc-4c6dbd111852.png)

https://user-images.githubusercontent.com/85202273/168954597-6ed89b11-d3cc-4065-a656-9dbf9d87f121.mov

winnie368c committed:
Added mypy type annotations stats_domain.py

winnie368c committed:
Added mypy type annotations to stats_domain.py

winnie368c committed:
Adding mypy type annotations to stats_domain.py

winnie368c committed:
Adding mypy type annotations to stats_domain.py

winnie368c committed:
Adding mypy type annotations to stats_domain.py

winnie368c committed:
Adding mypy type annotations to stats_domain.py

winnie368c committed:
Adding mypy type annotations to stats_domain.py

winnie368c committed:
Resolved merge conflicts by incorporating incoming changes

winnie368c committed:
Merge remote-tracking branch 'upstream/develop' into annotation-changes

winnie368c committed:
Addressing review comments

winnie368c committed:
Addressing review comments

winnie368c committed:
Addressing review comments

winnie368c committed:
Merge remote-tracking branch 'upstream/develop' into annotation-changes

winnie368c committed:
Addressing review comments

winnie368c committed:
Addressed review comments

winnie368c committed:
Merge remote-tracking branch 'upstream/develop' into annotation-changes

winnie368c committed:
Addressed review comments

winnie368c committed:
Addressed review comments

winnie368c committed:
Addressed review comments

winnie368c committed:
Addressing review comments

winnie368c committed:
Addressed review comments

winnie368c committed:
Addressed review comments

winnie368c committed:
Addressing review comments

winnie368c committed:
Adding mypy type annotations to question_fetchers.py

winnie368c committed:
addressing review comments

winnie368c committed:
fixing stats_domain_test.py tests

winnie368c committed:
Addressing review comments

vojtechjelinek commented on core/domain/stats_domain.py:
```diff
@@ -63,13 +66,133 @@
+    exp_id: str
+    exp_version: int
+    issue_type: str
+    # Type Dict[str, Dict[str, Any]] is used here because the type of the args
+    # value can differ.
```
Wrong comment?

winnie368c replied:
The default type of the customization arg is * in customization_args_util.py so it seems type of the args value can differ.

vojtechjelinek replied:
Yeah, but here we are not doing `Dict[str, Dict[str, Any]]` but `Dict[str, Any]`

winnie368c replied:
Oh I see sorry I missed that, done

vojtechjelinek commented on core/domain/stats_domain.py:
```diff
@@ -63,13 +66,133 @@
+    """Dictionary representing the LearnerAction object."""
+
+    action_type: str
+    # Type Dict[str, Dict[str, Any]] is used here because the type of the value
+    # of the args can differ.
```
Wrong comment?

winnie368c replied:
Done

vojtechjelinek commented on core/domain/stats_domain.py:
```diff
@@ -194,10 +322,16 @@ def get_sum_of_first_hit_counts(self):
+            'num_actual_starts_v2',
+            'num_completions_v1',
+            'num_completions_v2'
+        ]] = [
```
```suggestion
        exploration_stats_properties: List[Literal[
            'num_starts_v1',
            'num_starts_v2',
            'num_actual_starts_v1',
            'num_actual_starts_v2',
            'num_completions_v1',
            'num_completions_v2'
        ]] = [
```

winnie368c replied:
Done

vojtechjelinek commented on core/domain/stats_domain.py:
```diff
@@ -465,40 +610,41 @@ def __eq__(self, other):
-
-    def __hash__(self):
+        if not isinstance(other, StateStats):
+            # https://docs.python.org/2/library/constants.html#NotImplemented.
```
Why this leads to docs for Python 2.7?

vojtechjelinek replied:
ditto below

winnie368c replied:
Sorry that was an old comment - changed to Python 3.7 for NotImplemented docs

vojtechjelinek commented on core/domain/stats_domain.py:
```diff
@@ -746,8 +910,15 @@ class Playthrough:
+        exp_version: int,
+        issue_type: str,
+        # Type Dict[str, Dict[str, Any]] is used here because the type of the
+        # args value can differ.
```
Wrong comment?

winnie368c replied:
Done

vojtechjelinek commented on core/domain/stats_domain.py:
```diff
@@ -953,7 +1140,12 @@ def update_exp_issue_from_model(cls, issue_dict):
+    def _convert_issue_v1_dict_to_v2_dict(
+        cls,
+        issue_dict: Dict[
+            str, Union[str, Dict[str, Dict[str, str]], list[str], int, bool]
```
```suggestion
            str, Union[str, Dict[str, Dict[str, str]], List[str], int, bool]
```

winnie368c replied:
Done

vojtechjelinek reviewed (changes requested):
Thanks! Took a pass.

winnie368c committed:
Addressing review comments

winnie368c committed:
Merge remote-tracking branch 'upstream/develop' into annotation-changes

winnie368c committed:
Addressing review comments

winnie368c commented:
@vojtechjelinek PTAL

winnie368c committed:
fixing any comments

winnie368c commented:
@vojtechjelinek PTAL

seanlip reviewed (approved):
Thanks @winnie368c. LGTM for codeowners!

vojtechjelinek commented on core/domain/stats_domain.py:
```diff
@@ -63,13 +66,133 @@
+    classification_categorization: str
+    # Type Dict[str, Any] is used here because the type of the params values can
+    # differ.
+    params: Dict[str, Any]
```
Actually, one more thing, are we able to create a `Union` of allowed values instead of `Any` here?

vojtechjelinek replied:
And also elsewhere we use `Any`?

winnie368c replied:
Done

vojtechjelinek reviewed (changes requested):
Thanks! One more thing.

winnie368c committed:
fixing any

winnie368c committed:
Merge remote-tracking branch 'upstream/develop' into annotation-changes

winnie368c committed:
fixing any

winnie368c commented:
@vojtechjelinek PTAL

vojtechjelinek reviewed (approved):
LGTM!

vojtechjelinek commented:
@winnie368c The backend tests are failing, please try to fix it.

winnie368c committed:
fixing backend test failures

winnie368c committed:
Merge remote-tracking branch 'upstream/develop' into annotation-changes

winnie368c committed:
fixing backend test

winnie368c committed:
fixing backend tests

winnie368c committed:
fixing backend test

winnie368c committed:
fixing backend tests

winnie368c committed:
fixing keyerror backend tests

winnie368c committed:
fixing keyerror tests

winnie368c commented:
@vojtechjelinek The backend tests were failing because of the missing `answer_str` and `rule_spec_str` keys in the SubmittedAnswer dictionaries that we decided to type as Optional; I tried fixing them by adding them into the tests as `None` as I did in `stats_domain_test` back when those tests were failing as well. The backend tests on the dictionary seem to pass now but the backend check is still failing and I'm not sure why, could you please take a look?

vojtechjelinek commented:
@winnie368c The problem is that the _stats_domain_test.py_ doesn't fully cover _stats_domain.py_ 
![image](https://user-images.githubusercontent.com/10142938/173803873-56c4c5c2-7347-441b-984d-0b147cbde8a6.png)

You will need to add a test that covers line 1063.

winnie368c committed:
fixing backend test coverage

sahiljoster32 commented:
Hi @winnie368c , 
Can you please resolve the merge conflict by merging this PR with develop? -- Because I think your backend tests are now passing and hopefully after resolving the merge conflict it would be ready to merge into develop.

winnie368c committed:
resolving merge conflicts

winnie368c committed:
resolving merge conflicts

winnie368c committed:
Merge remote-tracking branch 'upstream/develop' into annotation-changes

winnie368c commented:
> Hi @winnie368c ,
> Can you please resolve the merge conflict by merging this PR with develop? -- Because I think your backend tests are now passing and hopefully after resolving the merge conflict it would be ready to merge into develop.

@sahiljoster32 I tried to resolve the merge conflict and push again but I'm not sure why the backend tests are failing now since I don't think I changed anything from the last push?

sahiljoster32 commented:
@winnie368c , Your e2e tests are also failing which is suspicious. Please merge your branch with develop. So that tests are able to re-run Thanks!!

sahiljoster32 commented:
Hi @winnie368c , 

I have gone through your `resolved merge conflict` commit and I found that you accidentally removed the whole `validate_aggregated_stats_dict` method and its corresponding test cases. So, that's why you are failing all these tests ( Actually that's what I suspect looking at your commits ). Please add the `validate_aggregated_stats_dict` method and its corresponding test in your code again.

winnie368c committed:
adding mypy type annotations for new methods

winnie368c committed:
adding mypy type annotations to validate_aggregated_stats_dict method

sahiljoster32 commented on core/domain/stats_domain.py:
```diff
@@ -806,8 +981,15 @@ class Playthrough:
+        exp_id: str,
+        exp_version: int,
+        issue_type: str,
+        issue_customization_args: Dict[str, Union[str, int, Dict[
+            Union[str, int], Union[str, int]
```
Ahh, I think @winnie368c you should change this type. Because looking at the test cases I think it should be `Dict[str, Dict[str, Union[str, int, ....]]` or something similar. For more info about `customization_args` you can read here. https://github.com/oppia/oppia/blob/2c89393c66636b399b93dc1397486c2f45e577f3/core/domain/customization_args_util.py#L84

sahiljoster32 replied:
Also, MyPy threw some errors like `unused 'type: ignore' comment` 
This means you should remove those ignores.

sahiljoster32 replied:
Apart from this everything looks good to me!!

winnie368c committed:
fixing customization_args type

winnie368c committed:
addressing review comments

sahiljoster32 commented:
@winnie368c I think you only fixed that place where I pointed!!!!. Can you please run the MyPy command locally (`python -m scripts.run_mypy_checks` ) to check all the errors, and it's a good practice to do that before committing the changes because in that way you don't have to wait for tests to be run on GitHub, right?

winnie368c commented:
> @winnie368c I think you only fixed that place where I pointed!!!!. Can you please run the MyPy command locally (`python -m scripts.run_mypy_checks` ) to check all the errors, and it's a good practice to do that before committing the changes because in that way you don't have to wait for tests to be run on GitHub, right?

Hi @sahiljoster32 , sorry I was going to reply to your comments from earlier but it looks like it didn't send! So I do run the mypy checks locally every time before pushing and I was actually confused because after deleting the `type: ignore[no-untyped-calls]`, mypy threw `no-untyped-call` errors again so I added them back, but on GitHub the tests are saying that these are unused.

sahiljoster32 commented:
@winnie368c if that's the case, you should first try to merge your branch with develop, because there are chances that your branch is behind the develop. 

There are 2 common good practices:
1. Always try to merge your branch with develop whenever you are committing new changes in your branch. So that both the branches are on same phase and your branch is always up-to-date.
2. Always run, tests locally if possible. This prevents delays in review process.

winnie368c committed:
Merge remote-tracking branch 'upstream/develop' into annotation-changes

winnie368c commented:
> @winnie368c if that's the case, you should first try to merge your branch with develop, because there are chances that your branch is behind the develop.
> 
> There are 2 common good practices:
> 
> 1. Always try to merge your branch with develop whenever you are committing new changes in your branch. So that both the branches are on same phase and your branch is always up-to-date.
> 2. Always run, tests locally if possible. This prevents delays in review process.

@sahiljoster32 Ah okay, totally forgot about merging but that makes a lot of sense and fixed the issue. I just have this one last question on an error for line 1404 of the test file. I changed the type of `action_customization_args` in the `LearnerAction` class to be` Dict[str, Dict[str, Any]]` since that is how it is typed in `customization_args_util.py`, but mypy is throwing this error. I have also tried `Dict[str, int]`, `Dict[str, Union[int, Dict[str, Any]]]`, and `Dict[str, Dict[str, Union[str, int]]`, all of which threw errors. Do you have any suggestions for what to type `action_customization_args`? Thanks.
![Screen Shot 2022-07-03 at 9 35 46 AM](https://user-images.githubusercontent.com/85202273/177053675-8d20bece-199c-45dc-aeff-a731c7a1d8b2.png)

sahiljoster32 commented:
> > @winnie368c if that's the case, you should first try to merge your branch with develop, because there are chances that your branch is behind the develop.
> > There are 2 common good practices:
> > 
> > 1. Always try to merge your branch with develop whenever you are committing new changes in your branch. So that both the branches are on same phase and your branch is always up-to-date.
> > 2. Always run, tests locally if possible. This prevents delays in review process.
> 
> @sahiljoster32 Ah okay, totally forgot about merging but that makes a lot of sense and fixed the issue. I just have this one last question on an error for line 1404 of the test file. I changed the type of `action_customization_args` in the `LearnerAction` class to be` Dict[str, Dict[str, Any]]` since that is how it is typed in `customization_args_util.py`, but mypy is throwing this error. I have also tried `Dict[str, int]`, `Dict[str, Union[int, Dict[str, Any]]]`, and `Dict[str, Dict[str, Union[str, int]]`, all of which threw errors. Do you have any suggestions for what to type `action_customization_args`? Thanks. ![Screen Shot 2022-07-03 at 9 35 46 AM](https://user-images.githubusercontent.com/85202273/177053675-8d20bece-199c-45dc-aeff-a731c7a1d8b2.png)

Okkkkhhhh, Actually this is wired because usually `customisation_args` defined under the format 
```python
key_name: {
    value: any_value
}
```
Ideally, the code you mentioned should have format like below ( as far as i know ): 
``` python
action_dict['action_customization_args']['new_key'] = {
    'value': 5
}
```
See if its working!! and if it works please make sure that you have changed the value of `new_key` in the test case, so that test cases does not fail  !!

@winnie368c I have replied to your comment!! PTAL!

sahiljoster32 commented:
Also, I think `Dict[str, Dict[str, Union[str, int]]]` is fine. Do not use `Dict[str, Dict[str, Any]]]` here!!.

winnie368c committed:
fix convert actionv1 to v2dict test

winnie368c committed:
fix convert action dict test

winnie368c committed:
Merge remote-tracking branch 'upstream/develop' into annotation-changes

winnie368c committed:
resolve merge conflicts

winnie368c committed:
fixed action dict conversion test

winnie368c committed:
fixed action dict conversion test

winnie368c commented:
@sahiljoster32 so sorry for the delay, I have made the final changes. Do you have any suggestions for how to fix `test_merge_conflict_present` failing for some reason in the backend tests? I merged my branch with develop and checked that there were no merge conflicts remaining; I am not sure what else to do. Thanks!

U8NWXD commented on scripts/linters/test_files/invalid_merge_conflict.py:
```diff
@@ -20,8 +20,4 @@
 
 from __future__ import annotations
 
-<<<<<<< HEAD
-'this is some content to mess with content to append.'
-=======
 'totally different content to merge later.'
->>>>>>> new_branch_to_merge_later
```
Why are you removing this? We have it here to test that our linter will flag unresolved merge conflicts

sahiljoster32 commented on core/domain/stats_domain.py:
```diff
@@ -63,13 +66,139 @@
 MAX_ANSWER_DETAILS_BYTE_SIZE = 10000
```
```suggestion
MAX_ANSWER_DETAILS_BYTE_SIZE: Final = 10000
```
Please add final to the lines where constants are defined (constants - which are defined with all capitals) 
Also, We use Final to tell MyPy that this value is not going to be changed and it is defined only for reading purpose.

sahiljoster32 commented on core/domain/stats_domain.py:
```diff
@@ -63,13 +66,139 @@
 MAX_ANSWER_DETAILS_BYTE_SIZE = 10000
```
```suggestion
MAX_ANSWER_DETAILS_BYTE_SIZE: Final = 10000
```
Please add final to the lines where constants are defined (constants - which are defined with all capitals) 
Also, We use Final to tell MyPy that this value is not going to be changed and it is defined only for reading purpose.

winnie368c replied:
Done

sahiljoster32 commented on core/domain/stats_domain.py:
```diff
@@ -1057,10 +1256,19 @@ def validate(self):
                     '%s' % type(playthrough_id))
 
 
+# In argument 'customization_args', we used Any type because it accepts the
+# values of customization args and that values can be of type str, int, Dict,
+# bool, List and other types too. So to make it generalize for every type of
+# values, we used Any here.
```
I think this comment should be removed. Because now we are not using Any, right?

winnie368c replied:
Done

sahiljoster32 reviewed (changes requested):
Hi @winnie368c , I have only a few comments PTAL!

sahiljoster32 commented on scripts/linters/test_files/invalid_merge_conflict.py:
```diff
@@ -20,8 +20,4 @@
 
 from __future__ import annotations
 
-<<<<<<< HEAD
-'this is some content to mess with content to append.'
-=======
 'totally different content to merge later.'
->>>>>>> new_branch_to_merge_later
```
@winnie368c Please revert these changes. The error you get in your PR is a backend flake.

winnie368c replied:
Done

winnie368c committed:
addressing review comments

winnie368c committed:
lint

winnie368c committed:
import final

sahiljoster32 commented:
@winnie368c Just one last thing did you update your branch with develop. Because there are some changes in develop that fixes the backend flake.

sahiljoster32 commented:
@winnie368c PLease merge your branch with develop once again. So that your PR is able to merge!

winnie368c commented:
> @winnie368c PLease merge your branch with develop once again. So that your PR is able to merge!

I merged earlier today before the last commit and now it shows working tree is clean, is this right
![Screen Shot 2022-07-16 at 3 27 37 PM](https://user-images.githubusercontent.com/85202273/179369393-24a0ff92-5e56-4020-a386-23d30e2ced32.png)
?

winnie368c committed:
Merge remote-tracking branch 'upstream/develop' into annotation-changes

sahiljoster32 reviewed (approved):
LGTM!!. All my comments are addressed, Thanks for the PR!

sahiljoster32 commented:
@U8NWXD PTAL!

U8NWXD reviewed (approved):
lgtm

winnie368c committed:
Merge remote-tracking branch 'upstream/develop' into annotation-changes

sahiljoster32 commented:
Hi @winnie368c Just one last thing, please update your branch with develop. So that we can merge this.

sahiljoster32 commented:
I'm merging this PR because all the comments are addressed and every reviewer approved the PR. Thanks @winnie368c !!
